addons/web/static/src/js/views/basic/basic_model.js

```diff
@@ -3653,6 +3653,7 @@ var BasicModel = AbstractModel.extend({
             active_ids: evalContext.id ? [evalContext.id] : [],
             active_model: element.model,
             current_date: moment().format('YYYY-MM-DD'),
+            current_datetime: moment().utc().format('YYYY-MM-DD HH:mm:ss'),
```

Comment:
Why `utc()`? Aren't date/time values supposed to be the browser time zone? The UTC stuff is only when exchanging datetime values with the server-side.

Also `current_datetime` is starting to be quite long, any problem with calling it `now` instead? Seems more readable and consistent with the rest of our code.
(`current_date` could probably be aliased to `today`, unless there's a problem with that?)

It doesn't seem likely that any model would have columns called `now` or `today` so collisions shouldn't be an issue.

Reply:
If I'm not mistaken the field value used in the [evaluation](https://github.com/odoo/odoo/blob/master/addons/web/static/lib/py.js/lib/py.js#L1344) comes unprocessed from the `fields_view_get` and is still formatted as a UTC date string. We could format both dates in the browser's time zone but that would be pointless since the value is only used for comparison purpose.

I like your idea for variable names: I think it is important to keep the information as simple and relevant as possible. I'll discuss it with the PO

Reply:
> If I'm not mistaken the field value used in the [evaluation](https://github.com/odoo/odoo/blob/master/addons/web/static/lib/py.js/lib/py.js#L1344) comes unprocessed from the `fields_view_get` and is still formatted as a UTC date string.

What does it have to do with `fields_view_get`? It only gives static metadata about the view.

The evaluation with pyeval is done at display time in the browser, and the value that you are comparing with (e.g. `date_planned_start`) directly comes from the data stored in the in-memory dataset, and it has already been translated to local time zone, no?

Also note that `current_date` was not converted to `utc()` above. You might think that dates are not affected by time zones, but they are too ;-)

We do treat dates differently because we don't offset them upon display (we consider that the date will be in the timezone of the person who set the field value), but comparisons still need to be done with the correct `current_date` as seen locally in the browser. If you see what I mean.

Reply:
> I'll discuss it with the PO

I suppose you meant "with @aab-odoo" ;-)

Reply:
> I double checked, and date(time) in the evaluation context are indeed UTC. Date(times) manipulated by the webclient are encoded into UTC moment instances, and they are converted into the user timezone at render time. But seeing how the evaluation context is [build](https://github.com/odoo/odoo/blob/master/addons/web/static/src/js/views/basic/basic_model.js#L3723), it would be UTC anyway, as "stringifying" a moment instance doesn't preverse the timezone.

Wait, how can we compare these things then?

```js
> // Assume local browser time is 2020-06-09 01:00:00 (Europe/Brussels)
> let today = moment('2020-06-09 01:00:00').format('YYYY-MM-DD')
> let now = moment('2020-06-09 01:00:00').utc().format('YYYY-MM-DD HH:MM:SS')
> let datefield_value = moment('2020-06-09 01:00:00').utc()
> let datefield_ctxval = JSON.parse(JSON.stringify(datefield_value))

> console.log(today)
2020-06-09
> console.log(now)
2020-06-08 23:06:00
> console.log(datefield_ctxval)
2020-06-08T23:00:00.000Z
```

Comparing lexicographically means the `T` in the stringified version would mess up everything? Or is it transformed again later before comparison?


Also note the problem with the value of `today` (aka `current_date`), which, if not forced to UTC, is not comparable with the context values, depending on the time. Hence my previous remark about consistency.

What am I missing? :thinking: 


> Shouldn't we add the datetime in [this context](https://github.com/odoo/odoo/blob/13.0/addons/web/static/src/js/core/py_utils.js#L256) as well (such that it can be used in, e.g., filter domains)?

Sure. But wait, why do we need 2 copies of these default context values? I thought these were the same thing.


> I also checked this one, and current_date is also the UTC date.

Well then that contradicts the docstring of `context_today()`, which corresponds to what I remembered about the evaluation of dates on the client side = local times ;-)
https://github.com/odoo/odoo/blob/13.0/addons/web/static/src/js/core/py_utils.js#L226-L236

We need to clarify this :)


> I agree about terms `now` and `today`. @Arcasias don't forget to keep `current_date` for backward compatibility (so `today` would only be an alias).

:+1:

Reply:
@odony 

> Comparing lexicographically means the `T` in the stringified version would mess up everything? Or is it transformed again later before comparison?
> 

Took me a while to understand, but [here](https://github.com/odoo/odoo/blob/master/addons/web/static/src/js/fields/field_utils.js#L436) it is... We override the toJSON function of moment instances we create for datetime field values, such that when they are sent to the server, they are properly formatted (without T/Z). As a (desired or not) side-effect, we get the same result when stringifying/parsing them for the eval context.

> Also note the problem with the value of `today` (aka `current_date`), which, if not forced to UTC, is not comparable with the context values, depending on the time. Hence my previous remark about consistency.

For dates, we indeed compare values in UTC with `current_date` in the user timezone. I think it's what we want. Dates are displayed in UTC (2020-06-11 in db is displayed 2020-06-11 in the client whatever the user timezone). Let's say we have a decoration-danger condition in a list, like `date_deadline >= today`. It is 1am, 2020-06-11, UTC. The user is in timezone UTC-4 (so 2020-06-10 9pm). There is a record with `date_deadline` being 2020-06-11. It's displayed 2020-06-11 to the user. If `current_date` was in UTC, the condition would be true and the row would be displayed in red (note than an hour earlier, it wouldn't have been). So I think what we have is correct, as the evaluation will be consistent with what's displayed.

And for datetimes, this doesn't apply as in this case, values are displayed in the user timezone, so we need to either compare UTC vs UTC, or user tz vs user tz.

Not sure I answered your question... If not (or if you disagree), would be easier to talk about it directly, but next week ^^

> Sure. But wait, why do we need 2 copies of these default context values? I thought these were the same thing.

Would be better indeed... We'll have a look.

Reply:
Hi @odony, so here is what we've got so far:
- the default context values are now held by `py_utils.pycontext()`
- this context is used by the `BasicModel` when assembling its own evaluation context
- this context holds the 2 new variables introduced by the task: `today` and `now`.
    - `today` is an alias of the `current_date` property which is now always formatted in the **local time zone** since they are expectedly set and compared in the browser time zone.
    - `now` is the current date and time value in **UTC** since the compared record values have yet to be converted to local time upon [display](https://github.com/odoo/odoo/blob/master/addons/web/static/src/js/fields/abstract_field.js#L390)

Please tell me if there is anything I'm missing

Reply:
@aab-odoo Thanks for the investigation and clarifications, everything you said makes sense :+1: 
To sum it up, we have:
- datetime: conversion UTC -> userTZ for display and comparison, value saved as UTC in db
- date: no conversion for display or comparison, value saved as userTZ in db

@Arcasias excellent work, it all sounds good :)